compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -0,0 +1,332 @@
+    if (auto blockArg = dyn_cast<BlockArgument>(l3SourceOffsetVal)) {
+      Operation *ownerOfBlockArg = blockArg.getOwner()->getParentOp();
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(blockArg.getOwner());
+      AffineExpr affineExpr = rewriter.getAffineDimExpr(0);
+      AffineExpr newAffineExpr = affineExpr + offsetToAdd;
+      auto newAffineMap = AffineMap::get(/*dimCount=*/1, /*symbolCount=*/0,
+                                         {newAffineExpr}, context);
+      newL3AsSourceOffset =
+          rewriter
+              .create<affine::AffineApplyOp>(ownerOfBlockArg->getLoc(),
+                                             newAffineMap, l3SourceOffsetVal)
+              .getResult();
+    } else {
+      Operation *defOpOfL3SourceOffset = l3SourceOffsetVal.getDefiningOp();
+      Location loc = defOpOfL3SourceOffset->getLoc();
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPoint(defOpOfL3SourceOffset);
+      if (auto applyOp =
+              dyn_cast<affine::AffineApplyOp>(defOpOfL3SourceOffset)) {
+        AffineExpr affineExpr = applyOp.getAffineMap().getResult(0);
+        AffineExpr newAffineExpr = affineExpr + offsetToAdd;
+        auto newAffineMap = AffineMap::get(
+            /*dimCount=*/1, /*symbolCount=*/0, {newAffineExpr}, context);
+        newL3AsSourceOffset =
+            rewriter
+                .create<affine::AffineApplyOp>(loc, newAffineMap,
+                                               applyOp.getMapOperands())
+                .getResult();
+      } else if (auto constantOffset = getConstantIntValue(l3SourceOffsetVal)) {
+        int64_t newOffset = *constantOffset + offsetToAdd;
+        newL3AsSourceOffset = rewriter.getIndexAttr(newOffset);
+      } else {
+        return failure();
+      }
+    }
+  }
+  return newL3AsSourceOffset;
+}
+
+LogicalResult splitLogicalObjectFifos(
+    IRRewriter &rewriter, SmallVector<AMDAIE::DmaCpyNdOp> &l2ToL1DmaOps,
+    MLIRContext *context) {
+  if (l2ToL1DmaOps.size() == 0) return success();
+
+  SmallVector<OpFoldResult> baseSourceOffsets =
+      l2ToL1DmaOps[0].getSourceMixedOffsets();
+  LogicalObjectFifoFromMemrefOp sourceObjectFifo =
+      l2ToL1DmaOps[0].getSourceObjectFifo();
+  auto sourceAllocOp =
+      sourceObjectFifo.getMemref().getDefiningOp<memref::AllocOp>();
+  if (!sourceAllocOp) {
+    sourceObjectFifo->emitOpError()
+        << "expected alloc op as the defining op of source "
+           "logicalobjectfifo.from_memref";
+    return failure();
+  }
+  // We will now capture those dimensions where L2 memory was split. The way we
+  // do this is by checking all L2->L1 DmaOps' source offset and marking those
+  // dimensions which are not equal to at least one of the source offsets.
+  DenseSet<unsigned> splitDimsSetForL2;
+  SmallVector<unsigned> splitDimsForL2;
+  for (unsigned i = 1, n = l2ToL1DmaOps.size(); i < n; i++) {
```

Comment:
We should check that all `l2ToL1DmaOps` have the same splitting dimensions.

Reply:
I don't think that is required.

Eg:  Assume the following bizarre case occurs :-
```
     A. L2 -> L1 : [0, 0, 0, 0]
     B. L2 -> L1 : [0, 0, 0, 0]
     C. L2 -> L1 : [0, 1, 0, 0]
     D. L2 -> L1 : [1, 0, 0, 0]
```
In this part of the code I'm performing a `any_of`-like operation i.e "is any of the value in the same dimension different".
So, in the above case the splitting dimensions for ALL L2->L1 ops would be dimension 0 and dimension 1 - I don't think a check is required.

And then in another utility I'm simply checking if they're contiguous.

Eg:  Let's take another bizarre case :-
```
     A. L2 -> L1 : [0, 0, 0, 1]
     B. L2 -> L1 : [0, 0, 0, 0]
     C. L2 -> L1 : [1, 1, 0, 0]
     D. L2 -> L1 : [0, 0, 0, 0]
```
Here dimension 0, 1 and 3 would make the splitting dimension.

But later in the verification of the splitting dimensions being contiguous it'd bail out successfully.

Could you explain why there is a need for the check (and how exactly) ?

Reply:
We should be prudent and bail out in these bizarre cases.

What about:

```
     A. L2 -> L1 : [0, 0, 1, 0]
     B. L2 -> L1 : [0, 0, 0, 0]
     C. L2 -> L1 : [1, 1, 0, 0]
     D. L2 -> L1 : [0, 0, 0, 0]
```

Here dimension 0, 1 and 2 would make the splitting dimension and dimensions are contiguous, but it's probably not going to go well.

> Could you explain why there is a need for the check (and how exactly) ?

We should be prudent and limit splitting buffers to the exact cases where we know it's going to work, i.e. all l2tol1dmas have the same splitting dimensions, contiguous range from zero, etc.

You can easily check it with a utility function that computes the dimensions for a single l2to1lDmaOp and then check that all of them have the same dimenions.